### PR TITLE
Loofah 14

### PR DESCRIPTION
**Pull request by vijamesDD**
- New collection in mongo called "categories" 
- API now accepts query for "allCategories" which returns categories and their associated skills within them
- allSkills endpoint remains unchanged
- I've done some refactoring to the naming of things to try remove reference to "skill" when it is not needed. 

Example request: 
```
localhost:8080/skills

{
    "query": "{allCategories{id, title, skills {id, title, description }}}"
}
```
Screen Shot 2020-07-07 at 4 16 49 PM

### CODE REVIEW

File: src/main/java/com/loofah/graph/api/services/SkillsService.java [hunk 1]
```diff
@@ -18,7 +20,11 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Map;
 
+//TODO online this is sometimes called the GraphQLProvider class instead of SkillService - perhaps should change the name?
+// https://www.graphql-java.com/tutorials/getting-started-with-spring-boot/
```
Comment (elishkaflint) [inline src/main/java/com/loofah/graph/api/services/SkillsService.java]: I think GraphQLProvider is much better than skills service, now that we have skills and categories 👍

File: src/main/java/com/loofah/graph/api/queries/AllCategoriesQuery.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        List<Skill> skills = skillsRepository.findAll();
+
+        return categories.stream()
+                .map(category -> {
+                    List<Skill> skillsInThisCategory = getSkillsFilteredByCategoryId(skills, category.getId());
+                    return new CategoryResponse(category.getId(), category.getTitle(), skillsInThisCategory);
+                }).collect(Collectors.toList());
+    }
+
+    private List<Skill> getSkillsFilteredByCategoryId(List<Skill> skills, int CategoryId) {
```
Comment (elishkaflint) [inline src/main/java/com/loofah/graph/api/queries/AllCategoriesQuery.java]: super nitpick but could we lowercase the CategoryId parameter?

File: src/main/java/com/loofah/graph/api/queries/AllCategoriesQuery.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Override
+    public List<CategoryResponse> get(DataFetchingEnvironment dataFetchingEnvironment) {
+        List<Category> categories = categoryRepository.findAll();
+        List<Skill> skills = skillsRepository.findAll();
+
+        return categories.stream()
+                .map(category -> {
+                    List<Skill> skillsInThisCategory = getSkillsFilteredByCategoryId(skills, category.getId());
+                    return new CategoryResponse(category.getId(), category.getTitle(), skillsInThisCategory);
```
Comment (elishkaflint) [inline src/main/java/com/loofah/graph/api/queries/AllCategoriesQuery.java]: been wondering about CategoryResponse. what do you think about making the name of this object a bit more specific like `CategoryWithSkills`?

Comment (elishkaflint) [reply src/main/java/com/loofah/graph/api/queries/AllCategoriesQuery.java]: And potentially the query could be called `AllCategoriesWithSkills`?

Comment (elishkaflint) [reply src/main/java/com/loofah/graph/api/queries/AllCategoriesQuery.java]: I'm definitely in two minds so let me know what you think!

File: src/main/java/com/loofah/graph/api/repositories/CategoryRepository.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.loofah.graph.api.repositories;
+
+import com.loofah.graph.api.models.database.Category;
+import org.springframework.data.mongodb.repository.MongoRepository;
+
+public interface CategoryRepository extends MongoRepository<Category, String> {
```
Comment (elishkaflint) [inline src/main/java/com/loofah/graph/api/repositories/CategoryRepository.java:6]: another **super** nitpick but we've got `SkillsRepository` and `CategoryRepository` - could we do Skills and Categories or Skill and Category?

File: src/main/java/com/loofah/graph/api/config/DatabaseSeeder.java [hunk 5]
```diff
@@ -10,17 +12,23 @@
...
     public void run(String... args) {
-        skillsRepository.save(Skill.builder().withId("1").withTitle("title1").withDescription( "description1").build());
-        skillsRepository.save(Skill.builder().withId("2").withTitle("title2").withDescription( "description2").build());
-        skillsRepository.save(Skill.builder().withId("3").withTitle("title3").withDescription( "description3").build());
+        skillsRepository.save(Skill.builder().withId("1").withTitle("title1").withDescription( "description1").withCategoryId(1).build());
+        skillsRepository.save(Skill.builder().withId("2").withTitle("title2").withDescription( "description2").withCategoryId(1).build());
+        skillsRepository.save(Skill.builder().withId("3").withTitle("title3").withDescription( "description3").withCategoryId(2).build());
+
+        categoryRepository.save(Category.builder().withId(1).withTitle("CategoryTitle1").build());
+        categoryRepository.save(Category.builder().withId(2).withTitle("CategoryTitle1").build());
```
Comment (elishkaflint) [inline src/main/java/com/loofah/graph/api/config/DatabaseSeeder.java]: can this be CategoryTitle2

File: src/main/java/com/loofah/graph/api/queries/AllCategoriesQuery.java (hunk 3 above)
Comment (vijamesDD) [reply src/main/java/com/loofah/graph/api/queries/AllCategoriesQuery.java]: 👍 Agreed to keep query name to AllCategories and update class to be CategoriesWithSkills. ✅

File: src/main/java/com/loofah/graph/api/services/SkillsService.java (hunk 1 above)
Comment (vijamesDD) [reply src/main/java/com/loofah/graph/api/services/SkillsService.java]: Agreed to: 
- Change name to GraphQLProvidor ✅ 
- Remove path '/skills' on our endpoint ✅ 
- refactor controller name from `skillsController` to just `Controller` ✅ 
- update tooling repo with correct queries now that endpoint is no longer /skills ✅  but not able to push yet due to permissions
- talk to Caitlin for FE / e2e test impact _I've done the changes myself to the frontend, https://github.com/Caitlin-cooling/loofah-frontend/pull/8_

**Review by Learner5200 — APPROVED** [review]

**Pull request merged**